policy: Allow non-standard scripts with -acceptnonstdtxn=1 (test nets only)

ajtowns opened the pull request:
This PR changes `-acceptnonstdtxn=1` so that it also skips the non-mandatory script checks, allowing txs that (eg) use OP_NOPx or OP_SUCCESS into the mempool. This remains only available on test nets.

ajtowns committed:
tests: in p2p_segwit, check non-mandatory errors with -acceptnonstdtxn=0 node

Prepare for updating -acceptnonstdtxn to allow txns that violate
STANDARD_SCRIPT_VERIFY_FLAGS but not MANDATORY_SCRIPT_VERIFY_FLAGS by
checking the non-mandatory flags with node that enforces standardness.

ajtowns committed:
validation: Check only (potential) block script flags when -acceptnonstdtxn is set

ajtowns commented:
Previous discussion in #28334.

If you enable this feature as a non-mining node, you risk having txs pinned in your mempool that conflict or double-spend standard txs in the normal mempool, which may cause problems with applications or layer-2 protocols.

If you enable this feature as a mining node, you have a small risk of ending up with invalid txs remaining in your mempool if a tx non-standard tx is accepted in your mempool and remains in there during a soft-fork activation that renders the tx invalid. At that point if that tx is included in blocks you generate, they will be invalid.

For these reasons this feature continues to only be available on test nets, not mainnet.

Having this be available in mainline bitcoin may make it easier to observe test transactions using proposed new soft-fork features (eg OP_CAT or SIGHASH_APO is enabled via the inquisition client which considers those txs standard, but the mainline client will consider them non-standard and has no option to accept them). [discussion link](https://twitter.com/glozow/status/1633134573356740608)

1440000bytes commented:
Concept ACK

benthecarman commented:
Concept ACK this would be nice for testing soft forks on test networks

luke-jr commented:
Seems like this goes beyond the expected behaviour. Maybe make it `=2` at least?

glozow commented:
concept ACK fwiw

instagibbs commented:
> Seems like this goes beyond the expected behaviour. Maybe make it =2 at least?

Since these are test-network-only changes, are we required to preserve current behavior?

ajtowns commented:
Bumped past #29086

luke-jr commented:
>Since these are test-network-only changes, are we required to preserve current behavior?

Whether you are required to or not, it is the correct thing to do. This option originated in Knots where it has always been available on mainnet, and expected to not enable upgradable opcodes/etc.

fjahr commented:
Concept ACK

ajtowns committed:
validation: Be explicit about whether CheckInputScripts is for block/mempool

Also moves the declaration for CheckInputScripts into the header to
avoid redeclaring it.

ajtowns committed:
tests: test tx rejection for both -acceptnonstdtxn=0 and 1

ismaelsadeeq reviewed (commented):
Concept ACK for the advantages stated.

1. I think this also makes it clear that a transaction input script must violate the mandatory script verify flags to be a consensus failure; other flags failure are just standardness.

I think the second commit adds some brittleness to the `CheckInputScripts` function; it now needs to know whether we are in block validation or individual transaction validation.
I think that should not be the case; the function should be as generic as possible, and each caller should know that context and hence use it as necessary. 
So `CheckInputScripts` should just return a validation state indicating whether consensus or standardness is violated; callers should add the detail of whether it is block validation or mempool acceptance validation that triggered the failure.

This makes the testing more straightforward, I think, and is better than the current smoke test.

See diff below.

2. I think after this PR, in non-block input script validation path, the result enum after a consensus input script failure is `TX_NOT_STANDARD`, which is incorrect; it should be `TX_CONSENSUS`. 

The diff below should fix this as well.


<details>
<summary>diff</summary>

```diff
diff --git a/src/test/txvalidationcache_tests.cpp b/src/test/txvalidationcache_tests.cpp
index c8cfbadfa9..9edad1d020 100644
--- a/src/test/txvalidationcache_tests.cpp
+++ b/src/test/txvalidationcache_tests.cpp
@@ -20,8 +20,8 @@ struct Dersig100Setup : public TestChain100Setup {
         : TestChain100Setup{ChainType::REGTEST, {.extra_args = {"-testactivationheight=dersig@102"}}} {}
 };

-bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
-                       const CCoinsViewCache& inputs, script_verify_flags flags, bool check_for_block,
+TxValidationState CheckInputScripts(const CTransaction& tx,
+                       const CCoinsViewCache& inputs, script_verify_flags flags,
                        bool cacheSigStore,
                        bool cacheFullScriptStore, PrecomputedTransactionData& txdata,
                        ValidationCache& validation_cache,
@@ -110,6 +110,21 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_block_doublespend, Dersig100Setup)
     BOOST_CHECK_EQUAL(m_node.mempool->size(), 0U);
 }

+void ValidTxValidationState(const TxValidationState& ret)
+{
+    BOOST_CHECK_EQUAL(ret.GetResult(), TxValidationResult::TX_RESULT_UNSET);
+    BOOST_CHECK(ret.GetRejectReason().empty());
+    BOOST_CHECK(ret.GetDebugMessage().empty());
+}
+
+void InvalidTxValidationState(const TxValidationState& ret, TxValidationResult result, const std::string& reject_reason, const std::string& debug_message)
+{
+    BOOST_CHECK(!ret.IsValid());
+    BOOST_CHECK_EQUAL(ret.GetResult(), result);
+    BOOST_CHECK_EQUAL(ret.GetRejectReason(), reject_reason);
+    BOOST_CHECK_EQUAL(ret.GetDebugMessage(), debug_message);
+}
+
 // Run CheckInputScripts (using CoinsTip()) on the given transaction, for all script
 // flags.  Test that CheckInputScripts passes for all flags that don't overlap with
 // the failing_flags argument, but otherwise fails.
@@ -128,8 +143,6 @@ static void ValidateCheckInputsForAllFlags(const CTransaction &tx, script_verify
     FastRandomContext insecure_rand(true);

     for (int count = 0; count < 10000; ++count) {
-        TxValidationState state;
-
         // Randomly selects flag combinations
         script_verify_flags test_flags = script_verify_flags::from_int(insecure_rand.randrange(MAX_SCRIPT_VERIFY_FLAGS));

@@ -143,23 +156,35 @@ static void ValidateCheckInputsForAllFlags(const CTransaction &tx, script_verify
             // WITNESS requires P2SH
             test_flags |= SCRIPT_VERIFY_P2SH;
         }
-        bool ret = CheckInputScripts(tx, state, &active_coins_tip, test_flags, /*check_for_block=*/false, true, add_to_cache, txdata, validation_cache, nullptr);
+        auto ret = CheckInputScripts(tx, &active_coins_tip, test_flags, true, add_to_cache, txdata, validation_cache, nullptr);
         // CheckInputScripts should succeed iff test_flags doesn't intersect with
         // failing_flags
         bool expected_return_value = !(test_flags & failing_flags);
-        BOOST_CHECK_EQUAL(ret, expected_return_value);
+        BOOST_CHECK_EQUAL(ret.IsValid(), expected_return_value);
+        if (!ret.IsValid()) {
+            if (test_flags & STANDARD_NOT_MANDATORY_VERIFY_FLAGS) {
+                BOOST_CHECK_EQUAL(ret.GetResult(), TxValidationResult::TX_NOT_STANDARD);
+            } else {
+                BOOST_CHECK_EQUAL(ret.GetResult(), TxValidationResult::TX_CONSENSUS);
+            }
+            BOOST_CHECK(!ret.GetRejectReason().empty());
+            BOOST_CHECK(!ret.GetDebugMessage().empty());
+        } else {
+           ValidTxValidationState(ret);
+        }

         // Test the caching
-        if (ret && add_to_cache) {
+        if (ret.IsValid() && add_to_cache) {
             // Check that we get a cache hit if the tx was valid
             std::vector<CScriptCheck> scriptchecks;
-            BOOST_CHECK(CheckInputScripts(tx, state, &active_coins_tip, test_flags, /*check_for_block=*/false, true, add_to_cache, txdata, validation_cache, &scriptchecks));
+            BOOST_CHECK(CheckInputScripts(tx, &active_coins_tip, test_flags, true, add_to_cache, txdata, validation_cache, &scriptchecks).IsValid());
             BOOST_CHECK(scriptchecks.empty());
         } else {
             // Check that we get script executions to check, if the transaction
             // was invalid, or we didn't add to cache.
             std::vector<CScriptCheck> scriptchecks;
-            BOOST_CHECK(CheckInputScripts(tx, state, &active_coins_tip, test_flags, /*check_for_block=*/false, true, add_to_cache, txdata, validation_cache, &scriptchecks));
+            auto cached_ret{CheckInputScripts(tx, &active_coins_tip, test_flags, true, add_to_cache, txdata, validation_cache, &scriptchecks)};
+            ValidTxValidationState(cached_ret);
             BOOST_CHECK_EQUAL(scriptchecks.size(), tx.vin.size());
         }
     }
@@ -214,16 +239,16 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, Dersig100Setup)
     {
         LOCK(cs_main);

-        TxValidationState state;
         PrecomputedTransactionData ptd_spend_tx;
-
-        BOOST_CHECK(!CheckInputScripts(CTransaction(spend_tx), state, &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_DERSIG, /*check_for_block=*/false, true, true, ptd_spend_tx, m_node.chainman->m_validation_cache, nullptr));
+        auto ret = CheckInputScripts(CTransaction(spend_tx), &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_DERSIG, true, true, ptd_spend_tx, m_node.chainman->m_validation_cache, nullptr);
+        InvalidTxValidationState(ret, TxValidationResult::TX_CONSENSUS, "Non-canonical DER signature", strprintf("input 0 of %s (wtxid %s), spending %s:0", spend_tx.GetHash().ToString(), CTransaction(spend_tx).GetWitnessHash().ToString(), m_coinbase_txns[0]->GetHash().ToString()));

         // If we call again asking for scriptchecks (as happens in
         // ConnectBlock), we should add a script check object for this -- we're
         // not caching invalidity (if that changes, delete this test case).
         std::vector<CScriptCheck> scriptchecks;
-        BOOST_CHECK(CheckInputScripts(CTransaction(spend_tx), state, &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_DERSIG, /*check_for_block=*/false, true, true, ptd_spend_tx, m_node.chainman->m_validation_cache, &scriptchecks));
+        auto non_cached_ret{CheckInputScripts(CTransaction(spend_tx), &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_DERSIG, true, true, ptd_spend_tx, m_node.chainman->m_validation_cache, &scriptchecks)};
+        ValidTxValidationState(non_cached_ret);
         BOOST_CHECK_EQUAL(scriptchecks.size(), 1U);

         // Test that CheckInputScripts returns true iff DERSIG-enforcing flags are
@@ -283,9 +308,9 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, Dersig100Setup)

         // Make it valid, and check again
         invalid_with_cltv_tx.vin[0].scriptSig = CScript() << vchSig << 100;
-        TxValidationState state;
         PrecomputedTransactionData txdata;
-        BOOST_CHECK(CheckInputScripts(CTransaction(invalid_with_cltv_tx), state, m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_CHECKLOCKTIMEVERIFY, /*check_for_block=*/false, true, true, txdata, m_node.chainman->m_validation_cache, nullptr));
+        auto valid_ret{CheckInputScripts(CTransaction(invalid_with_cltv_tx), m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_CHECKLOCKTIMEVERIFY, true, true, txdata, m_node.chainman->m_validation_cache, nullptr)};
+        ValidTxValidationState(valid_ret);
     }

     // TEST CHECKSEQUENCEVERIFY
@@ -311,9 +336,9 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, Dersig100Setup)

         // Make it valid, and check again
         invalid_with_csv_tx.vin[0].scriptSig = CScript() << vchSig << 100;
-        TxValidationState state;
         PrecomputedTransactionData txdata;
-        BOOST_CHECK(CheckInputScripts(CTransaction(invalid_with_csv_tx), state, &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_CHECKSEQUENCEVERIFY, /*check_for_block=*/false, true, true, txdata, m_node.chainman->m_validation_cache, nullptr));
+        auto valid_ret{CheckInputScripts(CTransaction(invalid_with_csv_tx), &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_CHECKSEQUENCEVERIFY, true, true, txdata, m_node.chainman->m_validation_cache, nullptr)};
+        ValidTxValidationState(valid_ret);
     }

     // TODO: add tests for remaining script flags
@@ -372,15 +397,16 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, Dersig100Setup)
         // Invalidate vin[1]
         tx.vin[1].scriptWitness.SetNull();

-        TxValidationState state;
         PrecomputedTransactionData txdata;
         // This transaction is now invalid under segwit, because of the second input.
-        BOOST_CHECK(!CheckInputScripts(CTransaction(tx), state, &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS, /*check_for_block=*/false, true, true, txdata, m_node.chainman->m_validation_cache, nullptr));
+        auto ret = CheckInputScripts(CTransaction(tx), &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS, true, true, txdata, m_node.chainman->m_validation_cache, nullptr);
+        InvalidTxValidationState(ret, TxValidationResult::TX_CONSENSUS, "Witness program hash mismatch", strprintf("input 1 of %s (wtxid %s), spending %s:1", tx.GetHash().ToString(), CTransaction(tx).GetWitnessHash().ToString(), spend_tx.GetHash().ToString()));

         std::vector<CScriptCheck> scriptchecks;
         // Make sure this transaction was not cached (ie because the first
         // input was valid)
-        BOOST_CHECK(CheckInputScripts(CTransaction(tx), state, &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS, /*check_for_block=*/false, true, true, txdata, m_node.chainman->m_validation_cache, &scriptchecks));
+        auto valid_ret{CheckInputScripts(CTransaction(tx), &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS, true, true, txdata, m_node.chainman->m_validation_cache, &scriptchecks)};
+        ValidTxValidationState(valid_ret);
         // Should get 2 script checks back -- caching is on a whole-transaction basis.
         BOOST_CHECK_EQUAL(scriptchecks.size(), 2U);
     }
diff --git a/src/validation.cpp b/src/validation.cpp
index 5d95edf4db..6c3644f9a5 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -138,8 +138,8 @@ const CBlockIndex* Chainstate::FindForkInGlobalIndex(const CBlockLocator& locato
     return m_chain.Genesis();
 }

-bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
-                       const CCoinsViewCache& inputs, script_verify_flags flags, bool check_for_block,
+TxValidationState CheckInputScripts(const CTransaction& tx,
+                       const CCoinsViewCache& inputs, script_verify_flags flags,
                        bool cacheSigStore,
                        bool cacheFullScriptStore, PrecomputedTransactionData& txdata,
                        ValidationCache& validation_cache,
@@ -429,7 +429,11 @@ static bool CheckInputsFromMempoolAndCache(const CTransaction& tx, TxValidationS
     }

     // Call CheckInputScripts() to cache signature and script validity against current tip consensus rules.
-    return CheckInputScripts(tx, state, view, flags, /*check_for_block=*/false, /* cacheSigStore= */ true, /* cacheFullScriptStore= */ true, txdata, validation_cache);
+    auto check_state = CheckInputScripts(tx, view, flags, /* cacheSigStore= */ true, /* cacheFullScriptStore= */ true, txdata, validation_cache);
+    if (!check_state.IsValid()) {
+        return state.Invalid(check_state.GetResult(), strprintf("mempool-script-verify-flag-failed (%s)", check_state.GetRejectReason()), check_state.GetDebugMessage());
+    }
+    return true;
 }

 namespace {
@@ -1253,13 +1257,15 @@ bool MemPoolAccept::PolicyScriptChecks(const ATMPArgs& args, Workspace& ws)

     // Check input scripts and signatures.
     // This is done last to help prevent CPU exhaustion denial-of-service attacks.
-    if (!CheckInputScripts(tx, state, m_view, scriptVerifyFlags, /*check_for_block=*/false, true, false, ws.m_precomputed_txdata, GetValidationCache())) {
+    TxValidationState check_state = CheckInputScripts(tx, m_view, scriptVerifyFlags, true, false, ws.m_precomputed_txdata, GetValidationCache());
+    if (!check_state.IsValid()) {
+        state.Invalid(check_state.GetResult(), strprintf("mempool-script-verify-flag-failed (%s)", check_state.GetRejectReason()), check_state.GetDebugMessage());
         // Detect a failure due to a missing witness so that p2p code can handle rejection caching appropriately.
         if (!tx.HasWitness() && SpendsNonAnchorWitnessProg(tx, m_view)) {
             state.Invalid(TxValidationResult::TX_WITNESS_STRIPPED,
                     state.GetRejectReason(), state.GetDebugMessage());
         }
-        return false; // state filled in by CheckInputScripts
+        return false;
     }

     return true;
@@ -2139,18 +2145,15 @@ ValidationCache::ValidationCache(const size_t script_execution_cache_bytes, cons
  *
  * Non-static (and redeclared) in src/test/txvalidationcache_tests.cpp
  */
-bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
-                       const CCoinsViewCache& inputs, script_verify_flags flags, bool check_for_block,
+TxValidationState CheckInputScripts(const CTransaction& tx,
+                       const CCoinsViewCache& inputs, script_verify_flags flags,
                        bool cacheSigStore,
                        bool cacheFullScriptStore, PrecomputedTransactionData& txdata,
                        ValidationCache& validation_cache,
                        std::vector<CScriptCheck>* pvChecks)
 {
-    if (tx.IsCoinBase()) return true;
-
-    if (flags & STANDARD_NOT_MANDATORY_VERIFY_FLAGS) {
-        assert(!check_for_block);
-    }
+    TxValidationState state;
+    if (tx.IsCoinBase()) return state;

     if (pvChecks) {
         pvChecks->reserve(tx.vin.size());
@@ -2166,7 +2169,7 @@ bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
     hasher.Write(UCharCast(tx.GetWitnessHash().begin()), 32).Write((unsigned char*)&flags, sizeof(flags)).Finalize(hashCacheEntry.begin());
     AssertLockHeld(cs_main); //TODO: Remove this requirement by making CuckooCache not require external locks
     if (validation_cache.m_script_execution_cache.contains(hashCacheEntry, !cacheFullScriptStore)) {
-        return true;
+        return state;
     }

     if (!txdata.m_spent_outputs_ready) {
@@ -2202,11 +2205,9 @@ bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
             // non-standard DER encodings or non-null dummy
             // arguments) or due to new consensus rules introduced in
             // soft forks.
-            if (!check_for_block) {
-                return state.Invalid(TxValidationResult::TX_NOT_STANDARD, strprintf("mempool-script-verify-flag-failed (%s)", ScriptErrorString(result->first)), result->second);
-            } else {
-                return state.Invalid(TxValidationResult::TX_CONSENSUS, strprintf("block-script-verify-flag-failed (%s)", ScriptErrorString(result->first)), result->second);
-            }
+            auto validation_result = flags & STANDARD_NOT_MANDATORY_VERIFY_FLAGS ? TxValidationResult::TX_NOT_STANDARD : TxValidationResult::TX_CONSENSUS;
+            state.Invalid(validation_result, ScriptErrorString(result->first), result->second);
+            return state;
         }
     }

@@ -2216,7 +2217,7 @@ bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
         validation_cache.m_script_execution_cache.insert(hashCacheEntry);
     }

-    return true;
+    return state;
 }

 bool FatalError(Notifications& notifications, BlockValidationState& state, const bilingual_str& message)
@@ -2663,21 +2664,20 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
         if (!tx.IsCoinBase() && fScriptChecks)
         {
             bool fCacheResults = fJustCheck; /* Don't cache results if we're actually connecting blocks (still consult the cache, though) */
-            bool tx_ok;
             TxValidationState tx_state;
             // If CheckInputScripts is called with a pointer to a checks vector, the resulting checks are appended to it. In that case
             // they need to be added to control which runs them asynchronously. Otherwise, CheckInputScripts runs the checks before returning.
             if (control) {
                 std::vector<CScriptCheck> vChecks;
-                tx_ok = CheckInputScripts(tx, tx_state, view, flags, /*check_for_block=*/true, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache, &vChecks);
-                if (tx_ok) control->Add(std::move(vChecks));
+                tx_state = CheckInputScripts(tx, view, flags, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache, &vChecks);
+                if (tx_state.IsValid()) control->Add(std::move(vChecks));
             } else {
-                tx_ok = CheckInputScripts(tx, tx_state, view, flags, /*check_for_block=*/true, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache);
+                tx_state = CheckInputScripts(tx, view, flags, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache);
             }
-            if (!tx_ok) {
+            if (!tx_state.IsValid()) {
                 // Any transaction validation failure in ConnectBlock is a block consensus failure
                 state.Invalid(BlockValidationResult::BLOCK_CONSENSUS,
-                              tx_state.GetRejectReason(), tx_state.GetDebugMessage());
+                              strprintf("block-script-verify-flag-failed (%s)", tx_state.GetRejectReason()), tx_state.GetDebugMessage());
                 break;
             }
         }

```
</details>

instagibbs reviewed (commented):
logic looks correct, some suggestions dfe4cc6cc2003b077f95e30e50c47f1fa28fb2e2

sedited commented on src/test/txvalidationcache_tests.cpp:
```diff
@@ -21,7 +21,8 @@ struct Dersig100Setup : public TestChain100Setup {
 
 bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
-                       const CCoinsViewCache& inputs, script_verify_flags flags, bool cacheSigStore,
+                       const CCoinsViewCache& inputs, script_verify_flags flags, bool check_for_block,
```
I have a slight preference for making this an enum, which seems more in line with communicating whether it is either block or mempool, like you state in the commit title. Something like


<details>
<summary>Click to expand diff</summary>

```diff
diff --git a/src/test/txvalidationcache_tests.cpp b/src/test/txvalidationcache_tests.cpp
index b7f90dd8b5..a4d53cd355 100644
--- a/src/test/txvalidationcache_tests.cpp
+++ b/src/test/txvalidationcache_tests.cpp
@@ -24 +24 @@ bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
-                       const CCoinsViewCache& inputs, script_verify_flags flags, bool check_for_block,
+                       const CCoinsViewCache& inputs, script_verify_flags flags, CheckType check_type,
@@ -146 +146 @@ static void ValidateCheckInputsForAllFlags(const CTransaction &tx, script_verify
-        bool ret = CheckInputScripts(tx, state, &active_coins_tip, test_flags, /*check_for_block=*/false, true, add_to_cache, txdata, validation_cache, nullptr);
+        bool ret = CheckInputScripts(tx, state, &active_coins_tip, test_flags, CheckType::Mempool, true, add_to_cache, txdata, validation_cache, nullptr);
@@ -156 +156 @@ static void ValidateCheckInputsForAllFlags(const CTransaction &tx, script_verify
-            BOOST_CHECK(CheckInputScripts(tx, state, &active_coins_tip, test_flags, /*check_for_block=*/false, true, add_to_cache, txdata, validation_cache, &scriptchecks));
+            BOOST_CHECK(CheckInputScripts(tx, state, &active_coins_tip, test_flags, CheckType::Mempool, true, add_to_cache, txdata, validation_cache, &scriptchecks));
@@ -162 +162 @@ static void ValidateCheckInputsForAllFlags(const CTransaction &tx, script_verify
-            BOOST_CHECK(CheckInputScripts(tx, state, &active_coins_tip, test_flags, /*check_for_block=*/false, true, add_to_cache, txdata, validation_cache, &scriptchecks));
+            BOOST_CHECK(CheckInputScripts(tx, state, &active_coins_tip, test_flags, CheckType::Mempool, true, add_to_cache, txdata, validation_cache, &scriptchecks));
@@ -220 +220 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, Dersig100Setup)
-        BOOST_CHECK(!CheckInputScripts(CTransaction(spend_tx), state, &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_DERSIG, /*check_for_block=*/false, true, true, ptd_spend_tx, m_node.chainman->m_validation_cache, nullptr));
+        BOOST_CHECK(!CheckInputScripts(CTransaction(spend_tx), state, &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_DERSIG, CheckType::Mempool, true, true, ptd_spend_tx, m_node.chainman->m_validation_cache, nullptr));
@@ -226 +226 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, Dersig100Setup)
-        BOOST_CHECK(CheckInputScripts(CTransaction(spend_tx), state, &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_DERSIG, /*check_for_block=*/false, true, true, ptd_spend_tx, m_node.chainman->m_validation_cache, &scriptchecks));
+        BOOST_CHECK(CheckInputScripts(CTransaction(spend_tx), state, &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_DERSIG, CheckType::Mempool, true, true, ptd_spend_tx, m_node.chainman->m_validation_cache, &scriptchecks));
@@ -288 +288 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, Dersig100Setup)
-        BOOST_CHECK(CheckInputScripts(CTransaction(invalid_with_cltv_tx), state, m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_CHECKLOCKTIMEVERIFY, /*check_for_block=*/false, true, true, txdata, m_node.chainman->m_validation_cache, nullptr));
+        BOOST_CHECK(CheckInputScripts(CTransaction(invalid_with_cltv_tx), state, m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_CHECKLOCKTIMEVERIFY, CheckType::Mempool, true, true, txdata, m_node.chainman->m_validation_cache, nullptr));
@@ -316 +316 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, Dersig100Setup)
-        BOOST_CHECK(CheckInputScripts(CTransaction(invalid_with_csv_tx), state, &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_CHECKSEQUENCEVERIFY, /*check_for_block=*/false, true, true, txdata, m_node.chainman->m_validation_cache, nullptr));
+        BOOST_CHECK(CheckInputScripts(CTransaction(invalid_with_csv_tx), state, &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_CHECKSEQUENCEVERIFY, CheckType::Mempool, true, true, txdata, m_node.chainman->m_validation_cache, nullptr));
@@ -378 +378 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, Dersig100Setup)
-        BOOST_CHECK(!CheckInputScripts(CTransaction(tx), state, &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS, /*check_for_block=*/false, true, true, txdata, m_node.chainman->m_validation_cache, nullptr));
+        BOOST_CHECK(!CheckInputScripts(CTransaction(tx), state, &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS, CheckType::Mempool, true, true, txdata, m_node.chainman->m_validation_cache, nullptr));
@@ -383 +383 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, Dersig100Setup)
-        BOOST_CHECK(CheckInputScripts(CTransaction(tx), state, &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS, /*check_for_block=*/false, true, true, txdata, m_node.chainman->m_validation_cache, &scriptchecks));
+        BOOST_CHECK(CheckInputScripts(CTransaction(tx), state, &m_node.chainman->ActiveChainstate().CoinsTip(), SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS, CheckType::Mempool, true, true, txdata, m_node.chainman->m_validation_cache, &scriptchecks));
diff --git a/src/validation.cpp b/src/validation.cpp
index 46a892d906..b0b77e3164 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -141 +141 @@ bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
-                       const CCoinsViewCache& inputs, script_verify_flags flags, bool check_for_block,
+                       const CCoinsViewCache& inputs, script_verify_flags flags, CheckType check_type,
@@ -431 +431 @@ static bool CheckInputsFromMempoolAndCache(const CTransaction& tx, TxValidationS
-    return CheckInputScripts(tx, state, view, flags, /*check_for_block=*/false, /* cacheSigStore= */ true, /* cacheFullScriptStore= */ true, txdata, validation_cache);
+    return CheckInputScripts(tx, state, view, flags, CheckType::Mempool, /* cacheSigStore= */ true, /* cacheFullScriptStore= */ true, txdata, validation_cache);
@@ -1147 +1147 @@ bool MemPoolAccept::PolicyScriptChecks(const ATMPArgs& args, Workspace& ws)
-    if (!CheckInputScripts(tx, state, m_view, scriptVerifyFlags, /*check_for_block=*/false, true, false, ws.m_precomputed_txdata, GetValidationCache())) {
+    if (!CheckInputScripts(tx, state, m_view, scriptVerifyFlags, CheckType::Mempool, true, false, ws.m_precomputed_txdata, GetValidationCache())) {
@@ -2063 +2063 @@ bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
-                       const CCoinsViewCache& inputs, script_verify_flags flags, bool check_for_block,
+                       const CCoinsViewCache& inputs, script_verify_flags flags, CheckType check_type,
@@ -2072 +2072 @@ bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
-        assert(!check_for_block);
+        assert(check_type == CheckType::Mempool);
@@ -2125 +2125 @@ bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
-            if (!check_for_block) {
+            if (check_type == CheckType::Mempool) {
@@ -2593 +2593 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
-                tx_ok = CheckInputScripts(tx, tx_state, view, flags, /*check_for_block=*/true, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache, &vChecks);
+                tx_ok = CheckInputScripts(tx, tx_state, view, flags, CheckType::Block, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache, &vChecks);
@@ -2596 +2596 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
-                tx_ok = CheckInputScripts(tx, tx_state, view, flags, /*check_for_block=*/true, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache);
+                tx_ok = CheckInputScripts(tx, tx_state, view, flags, CheckType::Block, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache);
diff --git a/src/validation.h b/src/validation.h
index cee0dd728b..6d8bd2ddab 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -259,0 +260,5 @@ struct PackageMempoolAcceptResult
+enum class CheckType {
+    Block,
+    Mempool
+};
+
```
</details>

ajtowns replied:
Made it an enum, but chose a slightly different name for it.

sedited reviewed (commented):
Approach ACK

Read through the discussion again, and think this is the pragmatic solution. Would also suggest removing the assert, as you mentioned above, and applying the other fixes from the review comments.

ajtowns commented on src/validation.cpp:
```diff
@@ -2257,6 +2257,10 @@ script_verify_flags GetBlockScriptFlags(const CBlockIndex& block_index, const Ch
 
+    // Note that any flags returned from this function (ie, specified
+    // here or in script_flag_exceptions) must also be included in
+    // MANDATORY_SCRIPT_VERIFY_FLAGS in policy/policy.h
```
I had a go previously at making this a compile time check, but it got kind-of ugly, especially when it's a no-op currently. Suggest exploring it further either in inquisition or when we have an actual deployment.

ajtowns replied:
Nevermind, I think this is done now.

ajtowns commented on test/functional/p2p_invalid_tx.py:
```diff
@@ -75,6 +80,11 @@ def run_test(self):
                 [tx], node, success=False,
                 reject_reason=template.reject_reason,
             )
+            std_reject = template.std_reject_reason or template.reject_reason
```
This change wouldn't be correct -- `BadInputOutpointIndex` tests something that's accepted as an orphan when relayed, but rejected when included in a block.

ajtowns committed:
validation: Ensure STANDARD_SCRIPT_VERIFY_FLAGS is a superset of GetBlockScriptFlags

ajtowns commented:
Updated. The approach I took was:

 * Introduced `GetAllConsensusScriptFlags()` that covers the superset of consensus flags that `GetBlockScriptFlags()` can return. This is evaluated at startup and stored in chainman.
 * `-acceptnonstdtxn` uses that value instead of `STANDARD_SCRIPT_VERIFY_FLAGS`. That way you won't accept any txs into the mempool that might be invalid via some deployment activation, even via a reorg. But if you disable a deployment (regtest or inquisition's `-renounce` feature), then you can override that restriction.
 * Compile-time check to ensure that `STANDARD_SCRIPT_VERIFY_FLAGS` covers everything that `GetBlockScriptFlags` might return -- this ensures you'll never have a consensus violating tx in your mempool in normal operation.
 * Runtime check to ensure that `GetAllConsensusScriptFlags()` covers everything that `GetBlockScriptFlags` does return -- this ensures you'll never have a consensus violating tx in your mempool with `-acceptnonstdtxn=1` either
 * `GetAllConsensusScriptFlags()` is in some sense duplicated code vs `GetBlockScriptFlags` -- it's essentially replacing `DeploymentActiveAt` with `DeploymentEnabled`.
 
 After this series of commits, `MANDATORY_SCRIPT_VERIFY_FLAGS` and `STANDARD_NOT_MANDATORY_VERIFY_FLAGS` are unused and could be dropped (well, MANDATORY merged into STANDARD, of course).
